src/cpp/ext/gsm/metadata_exchange.cc

```diff
@@ -178,59 +295,18 @@ ServiceMeshLabelsInjector::ServiceMeshLabelsInjector(
   // TODO(yashykt): Add mesh_id
 }
 
-std::vector<std::pair<absl::string_view, std::string>>
-ServiceMeshLabelsInjector::GetPeerLabels(
+std::unique_ptr<LabelsIterable> ServiceMeshLabelsInjector::GetPeerLabels(
     grpc_metadata_batch* incoming_initial_metadata) {
-  auto remote_metadata =
+  auto peer_metadata =
       incoming_initial_metadata->Take(grpc_core::XEnvoyPeerMetadata());
-  std::vector<std::pair<absl::string_view, std::string>> labels;
-  upb::Arena arena;
-  google_protobuf_Struct* struct_pb = nullptr;
-  if (remote_metadata.has_value()) {
-    std::string decoded_metadata;
-    bool metadata_decoded = absl::Base64Unescape(
-        remote_metadata.value().as_string_view(), &decoded_metadata);
-    if (metadata_decoded) {
-      struct_pb = google_protobuf_Struct_parse(
-          decoded_metadata.c_str(), decoded_metadata.size(), arena.ptr());
-    }
+  if (!peer_metadata.has_value()) {
+    return nullptr;
   }
-  labels.emplace_back(kPeerTypeAttribute,
-                      GetStringValueFromUpbStruct(
-                          struct_pb, kMetadataExchangeTypeKey, arena.ptr()));
-  labels.emplace_back(kPeerPodNameAttribute,
-                      GetStringValueFromUpbStruct(
-                          struct_pb, kMetadataExchangePodNameKey, arena.ptr()));
-  labels.emplace_back(
-      kPeerContainerNameAttribute,
-      GetStringValueFromUpbStruct(struct_pb, kMetadataExchangeContainerNameKey,
-                                  arena.ptr()));
-  labels.emplace_back(
-      kPeerNamespaceNameAttribute,
-      GetStringValueFromUpbStruct(struct_pb, kMetadataExchangeNamespaceNameKey,
-                                  arena.ptr()));
-  labels.emplace_back(
-      kPeerClusterNameAttribute,
-      GetStringValueFromUpbStruct(struct_pb, kMetadataExchangeClusterNameKey,
-                                  arena.ptr()));
-  labels.emplace_back(
-      kPeerLocationAttribute,
-      GetStringValueFromUpbStruct(struct_pb, kMetadataExchangeLocationKey,
-                                  arena.ptr()));
-  labels.emplace_back(
-      kPeerProjectIdAttribute,
-      GetStringValueFromUpbStruct(struct_pb, kMetadataExchangeProjectIdKey,
-                                  arena.ptr()));
-  labels.emplace_back(
-      kPeerCanonicalServiceAttribute,
-      GetStringValueFromUpbStruct(
-          struct_pb, kMetadataExchangeCanonicalServiceKey, arena.ptr()));
-  return labels;
+  return std::make_unique<PeerLabelsIterable>(std::move(peer_metadata).value());
```

Comment:
Suggest using `*std::move(peer_metadata)` instead of using the `value()` method.  See go/totw/181.

Reply:
ah, thanks! done